### rust in bioinformatics

A collection of genomics software tools written in Rust



### index section

##### bam
- [alignoth](https://github.com/alignoth/alignoth) : Creating alignment plots from bam files
- [bamrescue](https://github.com/arkanosis/bamrescue) : Utility to check Binary Sequence Alignment / Map (BAM) files for corruption and repair them
- [best](https://github.com/google/best) : Bam Error Stats Tool (best): analysis of error types in aligned reads
- [modkit](https://github.com/nanoporetech/modkit) : A bioinformatics tool for working with modified bases
- [mapAD](https://github.com/mpieva/mapAD) : An aDNA aware short-read mapper
- [perbase](https://github.com/sstadick/perbase) : Per-base per-nucleotide depth analysis
- [rustybam](https://github.com/mrvollger/rustybam) : bioinformatics toolkit in rust

##### csv

- [csview](https://github.com/wfxr/csview) : üì† Pretty and fast csv viewer for cli with cjk/emoji support
- [csvlens](https://github.com/YS-L/csvlens) : csvlens is a command line CSV file viewer. It is like less but made for CSV.
- [madato](https://github.com/inosion/madato) : Markdown Cmd Line, Rust and JS library for Excel to Markdown Tables
- [qsv](https://github.com/dathere/qsv) : Blazing-fast Data-Wrangling toolkit
- [rsv](https://github.com/ribbondz/rsv) : A command-line tool written in Rust for analyzing CSV, TXT, and Excel files.
- [tabiew](https://github.com/shshemi/tabiew) : A lightweight TUI app to view and query CSV files
- [tv](https://github.com/alexhallam/tv) : üì∫(tv) Tidy Viewer is a cross-platform CLI csv pretty printer that uses column styling to maximize viewer enjoyment.
- [xan](https://github.com/medialab/xan) : The CSV magician
- [xsv](https://github.com/BurntSushi/xsv) : A fast CSV command line toolkit written in Rust. ¬†
- [xtab](https://github.com/sharkLoc/xtab) : CSV command line utilities

##### dna

- [biotools](https://github.com/jimrybarski/biotools) : Command line bioinformatics functions
- [darwin](https://github.com/Ebedthan/darwin) : Create (rapid) neighbor-joining tree from sequences using mash distance
- [fakit](https://github.com/sharkLoc/fakit) : fakit: a simple program for fasta file manipulation
- [filterx](https://github.com/dwpeng/filterx) : process any file in tabular format. Fasta/fastq/GTF/GFF/VCF/SAM/BED
- [fq](https://github.com/stjude-rust-labs/fq) : Command line utility for manipulating Illumina-generated FASTQ files.
- [gsearch](https://github.com/jean-pierreboth/gsearch) : Approximate nearest neighbour search for microbial genomes based on hash metric
- [Hyper-Gen](https://github.com/wh-xu/Hyper-Gen) : HyGen: Compact and Efficient Genome Sketching using Hyperdimensional Vectors
- [kanpig](https://github.com/ACEnglish/kanpig) : Kmer Analysis of Pileups for Genotyping
- [kfc](https://github.com/lrobidou/KFC) : KFC (K-mer Fast Counter) is a fast and space-efficient k-mer counter based on hyper-k-mers.
- [ngs](https://github.com/stjude-rust-labs/ngs) : Command line utility for working with next-generation sequencing files.
- [nail](https://github.com/TravisWheelerLab/nail) : Nail is an Alignment Inference tooL
- [palindrome-finder](https://github.com/brianli314/palindrome-finder) : A bioinformatics tool written in Rust to find palindromic sequences in DNA
- [poasta](https://github.com/broadinstitute/poasta) : Fast and exact gap-affine partial order alignment
- [psdm](https://github.com/mbhall88/psdm) : Compute a pairwise SNP distance matrix from one or two alignment(s)
- [rust-bio-tools](https://github.com/rust-bio/rust-bio-tools) : A set of command line utilities based on Rust-Bio.
- [sigalign](https://github.com/baku4/sigalign) : A Similarity-Guided Alignment Algorithm
- [ska](https://github.com/bacpop/ska.rust) : Split k-mer analysis ‚Äì version 2
- [skc](https://github.com/mbhall88/skc) : Shared k-mer content between two genomes
- [sketchy](https://github.com/esteinig/sketchy) : Genomic neighbor typing of bacterial pathogens using MinHash üêÄ
- [tidk](https://github.com/tolkit/telomeric-identifier) : Identify and find telomeres, or telomeric repeats in a genome.
- [transanno](https://github.com/informationsea/transanno) : accurate LiftOver tool for new genome assemblies
- [xgt](https://github.com/Ebedthan/xgt) : Efficient and fast querying and parsing of GTDB's data

##### fastq

- [deacon](https://github.com/bede/deacon) : Fast (host) DNA sequence filtering 
- [fasten](https://github.com/lskatz/fasten) : üë∑ Fasten toolkit, for streaming operations on fastq files
- [faster](https://github.com/angelovangel/faster) :  A (very) fast program for getting statistics about a fastq file, the way I need them, written in Rust
- [fqgrep](https://github.com/fulcrumgenomics/fqgrep) : Grep for FASTQ files
- [fqkit](https://github.com/sharkLoc/fqkit) : ü¶Ä Fqkit: A simple and cross-platform program for fastq file manipulation ¬†
- [fqtk](https://github.com/fulcrumgenomics/fqtk) : Fast FASTQ sample demultiplexing in Rust.
- [grepq](https://github.com/rbfinch/grepq): quickly filter fastq files by matching sequences to a set of regex patterns
- [guide-counter](https://github.com/fulcrumgenomics/guide-counter) : A better, faster way to count guides in CRISPR screens.
- [K2Rmini](https://github.com/Malfoy/K2Rmini) : K2Rmini (or K-mer to Reads mini) is a tool to filter the reads contained in a FASTA/Q file based on a set of k-mers of interest.
- [kractor](https://github.com/sam-sims/kractor) : Rapidly extract reads from a FASTQ file based on taxonomic classification via Kraken2.
- [rasusa](https://github.com/mbhall88/rasusa) : Randomly subsample sequencing reads
- [SeqSizzle](https://github.com/ChangqingW/SeqSizzle) : SeqSizzle is a pager for viewing FASTQ files with fuzzy matching, allowing different adaptors to be colored differently.
- [sabreur](https://github.com/ebedthan/sabreur) : fast, reliable and handy demultiplexing tool for fastx files

#### format
- [atlas](https://github.com/stjude-rust-labs/atlas) : Enables storing, querying, transforming, and visualizing of multidimensional count data.
- [bigtools](https://github.com/jackh726/bigtools) : A high-performance BigWig and BigBed library in Rust
- [biotest](https://github.com/natir/biotest) : Generate random test data for bioinformatics
- [bqtools](https://github.com/arcinstitute/bqtools) : A command line utilty for working with BINSEQ files
- [cigzip](https://github.com/AndreaGuarracino/cigzip) : A tool for compression and decompression of alignment CIGAR strings using tracepoints.
- [d4tools](https://github.com/38/d4-format) : The D4 Quantitative Data Format
- [gfa2bin](https://github.com/MoinSebi/gfa2bin) : Convert various graph-related data to PLINK file. In addition, we offer multiple commands for filtering or modifying the generated PLINK files.
- [gia](https://github.com/noamteyssier/gia) : gia: Genomic Interval Arithmetic
- [granges](https://github.com/vsbuffalo/granges) : A Rust library and command line tool for working with genomic ranges and their data.
- [intspan](https://github.com/wang-q/intspan) : Command line tools for IntSpan related bioinformatics operations
- [nuc2bit](https://github.com/natir/nuc2bit) : A rust crate that provides methods for rapidly encoding and decoding nucleotides in 2-bit representation.
- [recmap](https://github.com/vsbuffalo/recmap) : A command line tool and Rust library for working with recombination maps.
- [transanno](https://github.com/informationsea/transanno) : accurate LiftOver tool for new genome assemblies
- [thirdkind](https://github.com/simonpenel/thirdkind) : Drawing reconciled phylogenetic trees allowing 1, 2 or 3 reconcillation levels
- [xsra](https://github.com/ArcInstitute/xsra) : An efficient CLI to extract sequences from the SRA


##### gff3

- [atg](https://github.com/anergictcell/atg) : A Rust library and CLI tool to handle genomic transcripts   
- [gffkit](https://github.com/sharkloc/gffkit) : a simple program for gff3 file manipulation

##### longreads

- [Autocycler](https://github.com/rrwick/Autocycler) : A tool for generating consensus long-read assemblies for bacterial genomes
- [chopper](https://github.com/wdecoster/chopper) : Rust implementation of [NanoFilt](https://github.com/wdecoster/nanofilt)+[NanoLyse](https://github.com/wdecoster/nanolyse), both originally written in Python. This tool, intended for long read sequencing such as PacBio or ONT, filters and trims a fastq file.
- [DeepChopper](https://github.com/ylab-hi/DeepChopper) : Language models identify chimeric artificial reads in NanoPore direct-RNA sequencing data.
- [fpa](https://github.com/natir/fpa) : Filter of Pairwise Alignement
- [herro](https://github.com/lbcb-sci/herro) : HERRO is a highly-accurate, haplotype-aware, deep-learning tool for error correction of Nanopore R10.4.1 or R9.4.1 reads (read length of >= 10 kbps is recommended).
- [HiPhase](https://github.com/PacificBiosciences/HiPhase) : Small variant, structural variant, and short tandem repeat phasing tool for PacBio HiFi reads
- [isONclust3](https://github.com/aljpetri/isONclust3) : De novo clustering of long transcript reads into genes
- [longshot](https://github.com/pjedge/longshot) : diploid SNV caller for error-prone reads
- [lrge](https://github.com/mbhall88/lrge) : Genome size estimation from long read overlaps
- [myloasm](https://github.com/bluenote-1577/myloasm) : A new high-resolution long-read metagenome assembler for even noisy reads
- [Polypolish](https://github.com/rrwick/Polypolish) : a short-read polishing tool for long-read assemblies
- [nextpolish2](https://github.com/Nextomics/NextPolish2) : Repeat-aware polishing genomes assembled using HiFi long reads
- [nanoq](https://github.com/esteinig/nanoq) : Minimal but speedy quality control for nanopore reads in Rust üêª
- [smrest](https://github.com/jts/smrest) : Tumour-only somatic mutation calling using long reads
- [trgt](https://github.com/PacificBiosciences/trgt) : Tandem repeat genotyping and visualization from PacBio HiFi data
- [yacrd](https://github.com/natir/yacrd) : Yet Another Chimeric Read Detector



##### metagenomics

- [coverm](https://github.com/wwood/CoverM) : Read coverage calculator for metagenomics
- [fairy](https://github.com/bluenote-1577/fairy) : alignment-free coverage calculation for metagenomic binning >100 times faster
- [floria](https://github.com/bluenote-1577/floria) : Strain-level haplotyping for metagenomes with short or long-reads.
- [galah](https://github.com/wwood/galah) : More scalable dereplication for metagenome assembled genomes
- [hyperex](https://github.com/Ebedthan/hyperex) : Hypervariable region primer-based extractor for 16S rRNA and other SSU/LSU sequences.
- [kun_peng](https://github.com/eric9n/Kun-peng) : Kun-peng: an ultra-fast, low-memory footprint and accurate taxonomy classifier for all
- [kmertools](https://github.com/anuradhawick/kmertools) : kmer based feature extraction tool for bioinformatics, metagenomics, AI/ML and more
- [kmerutils](https://github.com/jean-pierreBoth/kmerutils) : Kmer generating, counting hashing and related
- [Lorikeet](https://github.com/rhysnewell/Lorikeet) : Strain resolver for metagenomics
- [nohuman](https://github.com/mbhall88/nohuman) : Remove human reads from a sequencing run
- [rosella](https://github.com/rhysnewell/rosella) : Metagenomic Binning Algorithm
- [skani](https://github.com/bluenote-1577/skani) : Fast, robust ANI and aligned fraction for (metagenomic) genomes and contigs.
- [sourmash](https://github.com/sourmash-bio/sourmash) : Quickly search, compare, and analyze genomic and metagenomic data sets.
- [sylph](https://github.com/bluenote-1577/sylph) : ultrafast genome querying and taxonomic profiling for metagenomic samples by abundance-corrected minhash.
- [vircov](https://github.com/esteinig/vircov) : Viral genome coverage evaluation for metagenomic diagnostics ü©∏

##### pangenomics

- [impg](https://github.com/pangenome/impg) : implicit pangenome graph
- [panacus](https://github.com/marschall-lab/panacus) : Panacus is a tool for computing statistics for GFA-formatted pangenome graphs

##### phylogenomics

- [nextclade](https://github.com/nextstrain/nextclade) : Viral genome alignment, mutation calling, clade assignment, quality checks and phylogenetic placement
- [nwr](https://github.com/wang-q/nwr) : nwr is a command line tool for working with NCBI taxonomy, Newick files and assembly reports
- [unicore](https://github.com/steineggerlab/unicore) : Universal and efficient core gene phylogeny with Foldseek and ProstT5 
- [segul](https://github.com/hhandika/segul) : An ultrafast and memory efficient tool for phylogenomics

##### methylation
- [metheor](https://github.com/dohlee/metheor) : ‚òÑÔ∏è Ultrafast DNA methylation heterogeneity calculation from bisulfite alignments (Lee et al., PLOS Computational Biology. 2023)


##### proteomics

- [align-cli](https://github.com/snijderlab/align-cli) : A CLI for pairwise alignment of sequences, using both normal and mass based alignment.
- [daedalus](https://github.com/David-OConnor/daedalus) : Protein and molecule viewer
- [folddisco](https://github.com/steineggerlab/folddisco) : Fast indexing and search of discontinuous motifs in protein structures
- [foldmason](https://github.com/steineggerlab/foldmason) : Foldmason builds multiple alignments of large structure sets.
- [sage](https://github.com/lazear/sage) : Proteomics search & quantification so fast that it feels like magic

##### rna
- [oarfish](https://github.com/COMBINE-lab/oarfish) : long read RNA-seq quantification
- [rnapkin](https://github.com/ukmrs/rnapkin) : drawing RNA secondary structure with style; instantly
- [R2Dtool](https://github.com/comprna/R2Dtool) : R2Dtool is a set of genomics utilities for handling, integrating, and viualising isoform-mapped RNA feature data.
- [squab](https://github.com/zaeleus/squab) : Alignment-based gene expression quantification

##### singlecell

- [adview](https://github.com/JianYang-Lab/adview) : Adata Viewer: Head/Less/Shape h5ad file in terminal
- [alevin-fry](https://github.com/COMBINE-lab/alevin-fry) : üêü üî¨ü¶Ä alevin-fry is an efficient and flexible tool for processing single-cell sequencing data, currently focused on single-cell transcriptomics and feature barcoding.
- [cellranger](https://github.com/10XGenomics/cellranger) : 10x Genomics Single Cell Analysis
- [fragtk](https://github.com/stuart-lab/fragtk) : fragment file toolkit
- [precellar](https://github.com/regulatory-genomics/precellar) : Single-cell genomics preprocessing package
- [proseg](https://github.com/dcjones/proseg) : Probabilistic cell segmentation for in situ spatial transcriptomics
- [SnapATAC2](https://github.com/kaizhang/SnapATAC2) : Single-cell epigenomics analysis tools
- [spars](https://github.com/stuart-lab/spars) : Disk-based sparse matrix stats


##### slurm

- [ssubmit](https://github.com/mbhall88/ssubmit) : Submit slurm sbatch jobs without the need to create a script
- [wdl](https://github.com/stjude-rust-labs/wdl) : Rust crates for working with Workflow Description Language (WDL) documents.


##### vcf

- [echtvar](https://github.com/brentp/echtvar) : using all the bits for echt rapid variant annotation and filtering
- [gvcf_norm](https://github.com/mlin/gvcf_norm) : gVCF allele normalizer
- [mehari](https://github.com/varfish-org/mehari): VEP-like tool for sequence ontology and HGVS annotation of VCF files
- [vcf2parquet](https://github.com/natir/vcf2parquet) : Convert vcf in parquet
- [vcfexpress](https://github.com/brentp/vcfexpress) : expressions on VCFs
- [vcf-reformatter](https://github.com/flalom/vcf-reformatter) : üß¨ High-performance VCF file parser and reformatter with VEP annotation support. Converts complex VCF files to analyzable TSV format with intelligent transcript handling.


##### Gui

- [plascad](https://github.com/David-OConnor/plascad) : Design software for plasmid (vector) and primer creation and validation. Edit plasmids, perform PCR-based cloning, digest and ligate DNA fragments, and display details about expressed proteins. Integrates with online resources like NCBI and PDB.

##### other
- [biobear](https://github.com/wheretrue/biobear) : Work with bioinformatic files using Arrow, Polars, and/or DuckDB
- [binseq](https://github.com/ArcInstitute/binseq) : A high efficiency binary format for sequencing data
- [exon](https://github.com/wheretrue/exon) : Exon is an OLAP query engine specifically for biology and life science applications.
- [ggetrs](https://github.com/noamteyssier/ggetrs) : Efficient querying of biological databases
- [htsget-rs](https://github.com/umccr/htsget-rs) : A server implementation of the htsget protocol for bioinformatics in Rust
- [ibu](https://github.com/noamteyssier/ibu) : a rust library for high throughput binary encoding of genomic sequences
- [scidataflow](https://github.com/vsbuffalo/scidataflow): Command line scientific data management tool
- [sufr](https://github.com/TravisWheelerLab/sufr) : Parallel Construction of Suffix Arrays in Rust




## Starchart
<img src="https://starchart.cc/sharkLoc/rust-in-bioinformatics.svg" alt="Stargazers over time" style="max-width: 100%">
